### PR TITLE
Evaluators check for empty data

### PR DESCRIPTION
**Pull request by tovbinm**
**Related issues**
Most of our evaluators never check for empty during `evaluate` call and fail miserably.

**Describe the proposed solution**
- Added checks for empty data in `Evaluators` factory, which return defaults in case of no data was provided
- DRY out `Evaluators` factory 
- Update `EvaluatorsTest` to use property based testing with Tables. Each evaluators is now being evaluated against non empty and an empty dataset.
- Fix evaluation metrics order in ModelInsights pretty print
- Camel case Brier Score from `brierScore` -> `BrierScore`

**Describe alternatives you've considered**
NA

**Additional context**
Follows up on PR #176

### CODE REVIEW

File: core/src/main/scala/com/salesforce/op/evaluators/EvaluationMetrics.scala [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+   * Convert metrics class to a map
+   * @return a map from metric name to metric value
+   */
+  def toMap: Map[String, Any] = JsonUtils.toMap(JsonUtils.toJsonTree(this))
+
+  /**
+   * Convert metrics into metadata for saving
+   * @return metadata
+   */
+  def toMetadata: Metadata = this.toMap.toMetadata
```
Comment (leahmcguire) [inline core/src/main/scala/com/salesforce/op/evaluators/EvaluationMetrics.scala]: I think we have a trait for this ...

File: core/src/main/scala/com/salesforce/op/evaluators/Evaluators.scala [hunk 2]
```diff
@@ -85,8 +86,8 @@ object Evaluators {
         override def evaluate(dataset: Dataset[_]): Double = {
           import dataset.sparkSession.implicits._
           val dataUse = makeDataToUse(dataset, getLabelCol)
-          new MulticlassMetrics(dataUse.select(getPredictionValueCol, getLabelCol).as[(Double, Double)].rdd)
-            .precision(1.0)
+            .select(getPredictionValueCol, getLabelCol).as[(Double, Double)].rdd
+          if (dataUse.isEmpty()) 0.0 else new MulticlassMetrics(dataUse).precision(1.0)
```
Comment (leahmcguire) [inline core/src/main/scala/com/salesforce/op/evaluators/Evaluators.scala]: you pass this in in other places, why hardcode here?

File: core/src/main/scala/com/salesforce/op/evaluators/OpBinaryClassificationEvaluator.scala [hunk 3]
```diff
@@ -122,24 +122,32 @@ private[op] class OpBinaryClassificationEvaluator
...
+    default: => Double
+  ): Double = {
     val labelColName = getLabelCol
     val dataUse = makeDataToUse(dataset, labelColName)
     new BinaryClassificationEvaluator()
       .setLabelCol(labelColName)
       .setRawPredictionCol(getRawPredictionCol)
       .setMetricName(metricName.sparkEntryName)
-      .evaluate(dataUse)
+      .evaluateOrDefault(dataUse, default = default)
```
Comment (leahmcguire) [inline core/src/main/scala/com/salesforce/op/evaluators/OpBinaryClassificationEvaluator.scala]: We should not change the interface - the default has a default value so why change the name??

File: core/src/main/scala/com/salesforce/op/evaluators/EvaluationMetrics.scala (hunk 1 above)
Comment (tovbinm) [reply core/src/main/scala/com/salesforce/op/evaluators/EvaluationMetrics.scala]: yes, this exactly what I am using here - enrichment for map - https://github.com/salesforce/TransmogrifAI/blob/master/utils/src/main/scala/com/salesforce/op/utils/spark/RichMetadata.scala#L204

File: core/src/main/scala/com/salesforce/op/evaluators/OpBinaryClassificationEvaluator.scala (hunk 3 above)
Comment (tovbinm) [reply core/src/main/scala/com/salesforce/op/evaluators/OpBinaryClassificationEvaluator.scala]: it's an enrichment function for spark `Evaluator`

File: core/src/main/scala/com/salesforce/op/evaluators/Evaluators.scala (hunk 2 above)
Comment (tovbinm) [reply core/src/main/scala/com/salesforce/op/evaluators/Evaluators.scala]: Fixed.

**Pull request merged**